examples-gallery/plot_ball_rolling_on_spinning_disc.py

```diff
@@ -0,0 +1,620 @@
+
+# %%
+# The holonomic constraints must be fullfilled by the state constraints.
+x_start = 7.0
+q2_start = x_start / par_map[r]
+y_start = 7.0
+q1_start = -y_start / par_map[r]
+
+initial_state_constraints = {
+    q1: q1_start,
+    q2: q2_start,
+    q3: 0.0,
+    u1: 0.0,
+    u2: 0.0,
+    u3: 0.0,
+    x: x_start,
+    y: y_start,
+    ux: 0.0,
+    uy: 0.0
+    }
+final_state_constraints = {
+    x: 0.0,
+    y: 0.0,
+    ux: 0.0,
+    uy: 0.0,
+    }
+
+instance_constraints = (
+) + tuple(
+    xi.subs({t: t0}) - xi_val for xi, xi_val in
+    initial_state_constraints.items()
+) + tuple(
+    xi.subs({t: tf}) - xi_val for xi, xi_val in
+    final_state_constraints.items()
+)
+
+# %%
+# Forcing h > 0. helps to avoid negative h solutions.
+
+grenze = 10.0
+bounds = {t1: (-grenze, grenze), t2: (-grenze, grenze),
+    t3: (-grenze, grenze), h: (0.0001, 1.0)}
+
+# %%
+# Create an optimization problem.
+prob = Problem(
+    obj,
+    obj_grad,
+    eom,
+    state_symbols,
+    num_nodes,
+    interval_value,
+    known_parameter_map=par_map,
+    known_trajectory_map = {T(t): interval_fix},
+    instance_constraints=instance_constraints,
+    time_symbol=t,
+    bounds=bounds,
+    )
+
+# %%
+# The initial guess should meet the configuration constrains. It will be
+# plotted below.
+i1b = np.zeros(num_nodes)
+i2 = np.linspace(initial_state_constraints[x],
+    final_state_constraints[x], num_nodes)
+i1a = i2 / par_map[r]
+i3 = np.linspace(initial_state_constraints[y],
+    final_state_constraints[y], num_nodes)
+i1 = -i3 / par_map[r]
+i4 = np.zeros(8*num_nodes)
+initial_guess = np.hstack((i1,i1a, i1b, i2, i3, i4, 0.01))
+fig1, ax1 = plt.subplots(14, 1, figsize=(7.25, 1.25*14), sharex=True)
+prob.plot_trajectories(initial_guess, ax1)
+
+# %%
+# This way the maximum number ofinterations may be changed.
+# Default is 3000.
+prob.add_option('max_iter', 2000)
+
+# %%
+# Find the optimal solution.
+
+solution, info = prob.solve(initial_guess)
+print('message from optimizer:', info['status_msg'])
+print('Iterations needed',len(prob.obj_value))
+print(f'Optimal h = {solution[-1]:.3e}')
+
+## %%
+# PLot the objective value.
+prob.plot_objective_value()
+
+# %%
+# Plot the accuracy of the results.
+prob.plot_constraint_violations(solution)
+
+# %%
+# Plot the results.
+fig1, ax1 = plt.subplots(14, 1, figsize=(7.25, 3.25*14), sharex=True)
+prob.plot_trajectories(solution, ax1)
+
+# %%
+# Set up the EOMs to find the reaction forces.
+
+t = me.dynamicsymbols._t
+q1, q2, q3 = me.dynamicsymbols('q1 q2 q3')
+u1, u2, u3 = me.dynamicsymbols('u1 u2 u3')
+t1, t2, t3 = me.dynamicsymbols('t1 t2 t3')
+T = sm.symbols('T', cls=sm.Function)
+Tdot, Tdotdot = sm.symbols('Tdot Tdotdot')
+
+x, y = me.dynamicsymbols('x y')
+ux, uy = me.dynamicsymbols('ux uy')
+aux = me.dynamicsymbols('auxx, auxy, auxz')
+F_r = me.dynamicsymbols('Fx, Fy, Fz')
+mb, mo, g, r = sm.symbols('mb, mo, g, r')
+Omega, alpha = sm.symbols('Omega, alpha')
+rhs = list(sm.symbols('rhs:5'))
```

Comment:
Wasn't all of this already defined at the beginning of the file?

Reply:
> Wasn't all of this already defined at the beginning of the file?

For opty I set up the EOMs **without** virtual speeds / reaction forces. To get the reaction forces, I simply form the EOMsd again, but now **with** vih virtual speeds /reaction forces.

Reply:
Can't you solve the EOMs once with auxilary speeds? It gives you both the normal EoMs and the extra equations for the noncontributing forces in one go. I don't think there is any need to do it twice.

Reply:
> Can't you solve the EOMs once with auxilary speeds? It gives you both the normal EoMs and the extra equations for the noncontributing forces in one go. I don't think there is any need to do it twice.

I thought about this, but since opty overwrites the symbols, I thought it easier to do it twice. I could try to store the symbols like I did in another example.

Reply:
Can you give an example of opty overwriting symbols. If it does this, it is a bug.

Reply:
> Can you give an example of opty overwriting symbols? If it does this, it is a bug.
At least I thought this was happening im my simulation of my block sliding on a road, where I optimize twice. I thought this was normal for opty, so I used deepcopy to safe the symbols. 
Since you say this would be a bug, let me study it more closely.

Reply:
> Can't you solve the EOMs once with auxilary speeds? It gives you both the normal EoMs and the extra equations for the noncontributing forces in one go. I don't think there is any need to do it twice.

Since my 'overwriting fear' was wrong, I tried this:
- eliminated the last 3 equations from fr + frstar (there are 3 reaction forces)
- set the reaction forces to zero in the remaining (fr + frstar) equations
- all else unchanged: eom =(kd.col_join(fr_frstar_reduced)
-                               eom = eom.col_join(config_constraints)

But opty raises a fatal error.
No idea, why.
Unless I made a stupid mistake (also likely with me!) the safe route seems to be to set up the EOMs twice.

Reply:
The easy way for me to help you is that you push your changes to this branch and then I can run the code and see the error you get.